# **ENGG680 - Introduction to Digital Engineering**
## *Lab Assignment 4: ML Applications - Vehicle Motion State Classification*

## Preliminary: Certificate of Work


*We, the undersigned, certify that this is our own work, which has been done expressly for this course, either without the assistance of any other party or where appropriate we have acknowledged the work of others. Further, we have read and understood the section in the university calendar on plagiarism/cheating/other academic misconduct and we are aware of the implications thereof. We request that the total mark for this assignment be distributed as follows among group members:*

|          | First Name | Last Name | Signature (Full Name, Date) | Hours | Contribution % |
|----------|------------|-----------|-----------------------------|-------|----------------|
| Member 1: | Michael | Shi | Michael Shi, 11-25-2024 | 8 | 25% |
| Member 2: | Hlayn | Toe Wai | Hlayn Toe Wai, 11-28-2024 | 8 | 25% |
| Member 3: | Sheinn | Min Thu | Sheinn Min Thu, 11-28-2024 | 8 | 25% |
| Member 4: | Hsu Hsu | Sandi | Hsu Hsu Sandi, 11-28-2024 | 8 | 25% |


---

## Section 1 - Motion Classification Using *All Features* Derived From Inertial Sensor Data (33 Marks)

---

### 1.1 Assignment datasets import (6 Marks)

---

In [1]:
# some library imports
import pandas as pd

First load the following datasets into dataframes. 
* *UrbanNav-HK_Whampoa-20210521_sensors.csv* to dataframe variable named *cv_set*. This dataset is to be used for model training and hyper parameter tuning. 
* *UrbanNav-HK_TST-20210517_sensors.csv* to dataframe variable named *test_set1*. This is first independent test dataset. 
* *2020-03-14-16-45-35.csv* to dataframe variable named *test_set2*. This is the second independent test dataset.

In [2]:
# (2 marks)
# TODO: Load cross validation/training dataset (UrbanNav-HK_Whampoa-20210521_sensors.csv) to 'cv_set' dataframe

# uncomment and compelete below
cv_set = pd.read_csv("UrbanNav-HK_Whampoa-20210521_sensors.csv")

# TODO: Load UrbanNav-HK_TST-20210517_sensors.csv to 'test_set1' dataframe
test_set1 = pd.read_csv("UrbanNav-HK_TST-20210517_sensors.csv")

# TODO: Load 2020-03-14-16-45-35.csv to 'test_set2' dataframe
test_set2 = pd.read_csv("2020-03-14-16-45-35.csv")

For the *cv_set*, look at the dataframe head and info, and print the list of all columns. The *test_set1* and *test_set2* have the same schema as *cv_set*, however, have different number of rows (samples). 

In [3]:
# TODO: Display first two rows of cv_set 
# (1 mark)
print(cv_set.head(2))

     ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0 -0.003929 -0.093081  9.789349 -0.002649  0.001406 -0.001391  9.797418   
1 -0.000966 -0.092417  9.789095 -0.001806  0.000103 -0.001943  9.798815   

      g_sum        ax_var        ay_var  ...     g_min    ax_max    ay_max  \
0  0.022554  8.046094e-07  1.319665e-07  ...  0.000003  0.001496  0.000544   
1  0.021731  1.032619e-06  1.073067e-07  ...  0.000006  0.001865  0.000615   

     az_max    gx_max    gy_max    gz_max     a_max     g_max  motion_state  
0  0.024992  0.000110  0.000075  0.000078  0.024997  0.000187          rest  
1  0.024828  0.000124  0.000098  0.000090  0.024829  0.000144          rest  

[2 rows x 33 columns]


In [4]:
# TODO: Load the info method of cv_set
# (1 mark)
cv_set.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1534 entries, 0 to 1533
Data columns (total 33 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ax_sum        1534 non-null   float64
 1   ay_sum        1534 non-null   float64
 2   az_sum        1534 non-null   float64
 3   gx_sum        1534 non-null   float64
 4   gy_sum        1534 non-null   float64
 5   gz_sum        1534 non-null   float64
 6   a_sum         1534 non-null   float64
 7   g_sum         1534 non-null   float64
 8   ax_var        1534 non-null   float64
 9   ay_var        1534 non-null   float64
 10  az_var        1534 non-null   float64
 11  gx_var        1534 non-null   float64
 12  gy_var        1534 non-null   float64
 13  gz_var        1534 non-null   float64
 14  a_var         1534 non-null   float64
 15  g_var         1534 non-null   float64
 16  ax_min        1534 non-null   float64
 17  ay_min        1534 non-null   float64
 18  az_min        1534 non-null 

In [5]:
# TODO: Display the number of rows and columns  of all three datasets
# (1 mark)
print(f"'cv_set' has {cv_set.shape[0]} rows and {cv_set.shape[1]} columns")
print(f"'test_set1' has {test_set1.shape[0]} rows and {test_set1.shape[1]} columns")
print(f"'test_set2' has {test_set2.shape[0]} rows and {test_set2.shape[1]} columns")

'cv_set' has 1534 rows and 33 columns
'test_set1' has 783 rows and 33 columns
'test_set2' has 297 rows and 33 columns


In [6]:
# TODO: Display all columns in cv_set 
# (1 mark)
print(cv_set.columns.tolist())

['ax_sum', 'ay_sum', 'az_sum', 'gx_sum', 'gy_sum', 'gz_sum', 'a_sum', 'g_sum', 'ax_var', 'ay_var', 'az_var', 'gx_var', 'gy_var', 'gz_var', 'a_var', 'g_var', 'ax_min', 'ay_min', 'az_min', 'gx_min', 'gy_min', 'gz_min', 'a_min', 'g_min', 'ax_max', 'ay_max', 'az_max', 'gx_max', 'gy_max', 'gz_max', 'a_max', 'g_max', 'motion_state']


In all three dataframes, target vector is stored in the *motion_state* column. All other columns are features. 

___

### 1.2 - Grid Search with Cross Validation (18 Marks)

___

#### 1.2.1 - Extract all features and target from cv_set

___

From the *cv_set*, load the target vector stored in the *motion_state* column into a new variable *y* 

In [7]:
# TODO - Load target vector into y
# (1 mark) 
y = cv_set['motion_state']

From *cv_set*, load all features, that is all columns except for the target vector, in a new dataframe variable *X*

In [8]:
# TODO - Load all features from cv_set to X
# (2 marks)
X = cv_set.drop(columns=['motion_state'])
X.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0 -0.003929 -0.093081  9.789349 -0.002649  0.001406 -0.001391  9.797418   
1 -0.000966 -0.092417  9.789095 -0.001806  0.000103 -0.001943  9.798815   

      g_sum        ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.022554  8.046094e-07  1.319665e-07  ...  0.023871  0.000003  0.001496   
1  0.021731  1.032619e-06  1.073067e-07  ...  0.024153  0.000006  0.001865   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.000544  0.024992  0.000110  0.000075  0.000078  0.024997  0.000187  
1  0.000615  0.024828  0.000124  0.000098  0.000090  0.024829  0.000144  

[2 rows x 32 columns]

___

#### 1.2.2 - Configure and Perform Grid Search with Cross-Validation

___

In [9]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler, RobustScaler, MinMaxScaler, Normalizer
from sklearn.model_selection import GridSearchCV

The dataset *(X, y)*, must be split into training/validation *(X_train, y_train)* and testing *(X_test, y_test)* sets. Use the `train_test_split` function from sklearn to create the split, where  **30%** of data is used for testing. Set the `random_state=42`, and `stratify=y`.  

In [10]:
# TODO - Perform test/train split 
# (1 mark)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.30, random_state=42, stratify=y
)
print(f"Training set: {X_train.shape[0]} samples, Testing set: {X_test.shape[0]} samples")

Training set: 1073 samples, Testing set: 461 samples


Initialize a `Pipeline`, consisting of a *scaler* and a *classifer*, set the default values to *StandardScaler()*, and *SVC()*, respectively. 

In [11]:
# TODO - Initialize Pipeline into variable pipe
# (2 marks)
pipe = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', SVC())
])

Define the grid search configuration (*param_grid*) for training and hyper parameter tuning of the *support vector classifier*. Tune, the hyper-parameter *C*, over the interval `C=[0.01, 0.1, 1 ,10,100]`, *gamma*, over the interval `gamma=[0.001, 0.01, 0.1,1]`, and use the *radial basis function* as the kernel. For the *scaler* parameter, test with the `StandardScaler`, `RobustScaler`, and `None` options. 

In [12]:
# TODO - Setup grid set 
# (4 marks)

param_grid = {
    'scaler': [StandardScaler(), RobustScaler(), None],
    'classifier__C': [0.01, 0.1, 1, 10, 100],
    'classifier__gamma': [0.001, 0.01, 0.1, 1],
    'classifier__kernel': ['rbf']  # Radial Basis Function
}

In [13]:
# Perform the grid search to train model and tune hyper parameters
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy', return_train_score=True)
grid.fit(X_train, y_train) 

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('classifier', SVC())]),
             param_grid={'classifier__C': [0.01, 0.1, 1, 10, 100],
                         'classifier__gamma': [0.001, 0.01, 0.1, 1],
                         'classifier__kernel': ['rbf'],
                         'scaler': [StandardScaler(), RobustScaler(), None]},
             return_train_score=True, scoring='accuracy')

___

#### 1.2.3 - Visualize Grid Search Results

___

Print the results from the grid search. The following outputs must be printed.
* Best parameter set
* Best cross-validation train score
* Best cross-validation test score
* Test set accuracy

In [14]:
# TODO: Print the results from the grid search
# (2 marks)

from sklearn.metrics import accuracy_score

print("Best params:\n{}\n".format(grid.best_params_))
print("Best cross-validation train score: {:.2f}".format(max(grid.cv_results_['mean_train_score'])))
print("Best cross-validation validation score: {:.2f}".format(grid.best_score_))

test_set_score = accuracy_score(y_test, grid.best_estimator_.predict(X_test))
print("Test-set score: {:.2f}".format(test_set_score))

Best params:
{'classifier__C': 100, 'classifier__gamma': 0.01, 'classifier__kernel': 'rbf', 'scaler': RobustScaler()}

Best cross-validation train score: 1.00
Best cross-validation validation score: 0.95
Test-set score: 0.96


Display the confusion matrix and print the classification report for the trained model. Use the *(X_test, y_test)* dataset. **Hint** - You can use your code from assignment 3.

In [15]:
# TODO - Display confusion matrix and print classification report for (X_test, y_test)
# (2 marks)

from sklearn.metrics import confusion_matrix, classification_report

y_pred = grid.best_estimator_.predict(X_test)

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Confusion Matrix:
[[153   2   0]
 [ 10 192   5]
 [  0   3  96]]

Classification Report:
              precision    recall  f1-score   support

        rest       0.94      0.99      0.96       155
    straight       0.97      0.93      0.95       207
        turn       0.95      0.97      0.96        99

    accuracy                           0.96       461
   macro avg       0.95      0.96      0.96       461
weighted avg       0.96      0.96      0.96       461



___

#### 1.2.4 - Questions

1. Comment on the fit of the model to the dataset? (2 marks)
2. In the grid search, are the intervals used for *C* and *gamma* hyperparameters suitable? Why or why not? (2 marks)

#### 1.2.4 - Answers
1. The model fits the dataset very well, achieving high accuracy (96%) on the test set and strong cross-validation performance (95% validation score). The best cross-validation train score is 1.00, indicating the model fits the training data almost perfectly. The classification report shows balanced precision, recall, and F1-scores: high precision (0.94–0.97), recall (0.93–0.99), and F1-scores (0.95–0.96) across all motion states (rest, straight, turn). The confusion matrix shows low misclassification rates, for example there are only 10 "straight" samples misclassified as "rest". This indicates minimal misclassification, good generalization, and minimal overfitting.

2. The intervals for C (0.01–100) and gamma (0.001–1) are suitable, as the grid search identified optimal parameters within these ranges (C=100 and gamma=0.01, both within the specified intervals). The chosen value of C (100) indicates the model prefers weaker regularization. The optimal value of gamma (0.01) suggests a moderate influence radius. Optimizing the ranges for these values could potentially further improve performance.

___

### 1.3 - Model Evaluation on Test Datasets 1 and 2 (9 Marks)

___

#### 1.3.1 - Extract all features and targets from test datasets

Next, the model trained in 1.2 will be tested on independent datasets, that is *test_set1* and *test_set2*. <br>
The test_sets must be seperated into features (Xs) and targets (ys).

In [16]:
# TODO - Load the target vector from test_set1 into y1, and all features from test_set1 into X1
# (1 marks)
y1 = test_set1['motion_state']
X1 = test_set1.drop(columns=['motion_state'])
X1.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0  0.264465  0.089904  9.786930 -0.002305  0.000218 -0.001143  9.801661   
1  0.264300  0.088882  9.788127 -0.001865 -0.000523 -0.001999  9.802481   

      g_sum    ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.025634  0.000001  7.749919e-08  ...  0.024056  0.000003  0.002506   
1  0.025218  0.000001  9.243000e-08  ...  0.024084  0.000007  0.002427   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.000845  0.024884  0.000155  0.000078  0.000069  0.024886  0.000170  
1  0.000949  0.024881  0.000201  0.000095  0.000076  0.024883  0.000201  

[2 rows x 32 columns]

In [17]:
# TODO - Load the target vector from test_set2 into y2, and all features from test_set2 into X2
y2 = test_set2['motion_state']
X2 = test_set2.drop(columns=['motion_state'])
X2.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0 -0.046400  0.305907  9.791451 -0.002184  0.001311 -0.001901  9.800683   
1 -0.050073  0.311947  9.785619 -0.003369  0.001898 -0.002051  9.797741   

      g_sum    ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.012437  0.000002  3.020959e-07  ...  0.048406  0.000007  0.002575   
1  0.013805  0.000003  6.859937e-07  ...  0.047696  0.000013  0.003880   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.002993  0.049701  0.000094  0.000115  0.000080  0.049764  0.000142  
1  0.004106  0.050477  0.000140  0.000140  0.000083  0.050763  0.000151  

[2 rows x 32 columns]

____

#### 1.3.2 - Visualize Results

For each test dataset, print the test set accuracy score. 

In [18]:
# TODO - Print test score accuracies for test_set1 and test_set2
# (1 marks)

test_set1_score = accuracy_score(y1, grid.best_estimator_.predict(X1))
test_set2_score = accuracy_score(y2, grid.best_estimator_.predict(X2))

print("test_set1 score: {:.2f}".format(test_set1_score))
print("test_set2 score: {:.2f}".format(test_set2_score))

test_set1 score: 0.92
test_set2 score: 0.25


Display the confusion matrix and print the classification for test_set 1 *(X1, y1)*

In [19]:
# TODO - Display confusion matrix and print classification report for (X1, y1)
# (1 marks)
y1_pred = grid.best_estimator_.predict(X1)

print("\nConfusion Matrix for test_set1:")
print(confusion_matrix(y1, y1_pred))

print("\nClassification Report for test_set1:")
print(classification_report(y1, y1_pred))


Confusion Matrix for test_set1:
[[200  31   0]
 [  7 387  17]
 [  0   7 134]]

Classification Report for test_set1:
              precision    recall  f1-score   support

        rest       0.97      0.87      0.91       231
    straight       0.91      0.94      0.93       411
        turn       0.89      0.95      0.92       141

    accuracy                           0.92       783
   macro avg       0.92      0.92      0.92       783
weighted avg       0.92      0.92      0.92       783



Display the confusion matrix and print the classification for test_set 2 *(X2, y2)*

In [20]:
# TODO - Display confusion matrix and print classification report for (X2, y2)
y2_pred = grid.best_estimator_.predict(X2)

print("\nConfusion Matrix for test_set2:")
print(confusion_matrix(y2, y2_pred))

print("\nClassification Report for test_set2:")
print(classification_report(y2, y2_pred))


Confusion Matrix for test_set2:
[[  0   0  16]
 [  0   0 206]
 [  0   0  75]]

Classification Report for test_set2:
              precision    recall  f1-score   support

        rest       0.00      0.00      0.00        16
    straight       0.00      0.00      0.00       206
        turn       0.25      1.00      0.40        75

    accuracy                           0.25       297
   macro avg       0.08      0.33      0.13       297
weighted avg       0.06      0.25      0.10       297



c:\Users\mjssh\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\mjssh\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\mjssh\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, mo

___

#### 1.3.3 - Questions 

**Based on results so far** 
1. Comment on the model's performances on the datasets *(X_test, y_test)*, *(X1, y1)*, *(X2, y2)*? A significant difference in performance should be observed, what is this an example of? (3 marks)
1. Provide and explain potential causes for the observed behaviour in performance? (3 marks)

#### 1.3.3 - Answers
1. The model performs well on the (X_test, y_test) dataset with an accuracy of 96%, it also performs well on the first test set (X1, y1) with an accuracy of 92%. However, on the second test set (X2, y2) the accuracy is poor at 25%, as it only able to predict the "turn" class correctly while failing to classify "rest" and "straight" entirely. This significant drop in performance on the second test set suggests domain mismatch or data shift, where the characteristics of the second test set differ significantly from those of the training and validation data.

2. Some potential causes for the observed behaviour in performance are:
    - Imbalanced Class Representation: The model struggles to classify "rest" and "straight" in the second test set, possibly because the training data doesn't include enough examples or differences in "rest" and "straight" that match the conditions in the second test set.
    - Feature Distribution Mismatch: The features in the second test set may have a distribution that is different from the training set and first test set.
    - Overfitting to Training/Validation Data: The model might have overfit the training data, including biases present in the training set and first test set. This would limit its ability to generalize to new datasets like the second test set.
    - Measurement Noise or Errors: The second test set might have noisy or corrupted data, which the model was not trained to handle, leading to misclassfication.

___

## Section 2 - Classification Using *Selected Features* Derived From Inertial Sensor Data (12 Marks)

In this second section, an *SVC* model will be trained and tuned using grid search, however, only *selected features* will be used.

___

### 2.1 Grid Search with Cross Validation (5 Marks)

___

#### 2.1.1 Extract *selected* features and target from *cv_set*

From the *cv_set* dataframe, extract the selected features, listed below, and load them in a new dataframe *X*.
* *ax_var* - variance of accelerometer readings in x
* *ay_var* - variance of accelerometer readings in y
* *az_sum* - sum of accelerometer z readings
* *a_sum* - sum of the norm of accelerometer x, y, z readings
* *gz_var* - variance of gyroscope reading in z

In [21]:
# TODO - Load selected features from cv_set to X
# (1 marks)
selected_features = ['ax_var', 'ay_var', 'az_sum', 'a_sum', 'gz_var']
X = cv_set[selected_features]
X.head(2)

ax_var        ay_var    az_sum     a_sum        gz_var
0  8.046094e-07  1.319665e-07  9.789349  9.797418  7.446916e-10
1  1.032619e-06  1.073067e-07  9.789095  9.798815  6.983481e-10

In [22]:
# TODO - Load motion_state column (target vector) into y
y = cv_set['motion_state']
y.head(2)

0    rest
1    rest
Name: motion_state, dtype: object

___

#### 2.1.2 - Configure and Perform Grid Search with Cross-Validation

Follow the steps from **1.2.2** and perform model training and tuning using the selected feature set.

In [23]:
# TODO - Perform model training and tuning using `GridSearchCV`

# Perform test/train split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.30, random_state=42, stratify=y
)
print(f"Training set: {X_train.shape[0]} samples, Testing set: {X_test.shape[0]} samples")

# Initialize Pipeline into variable pipe
pipe = Pipeline([
    ('scaler', StandardScaler()),
    ('classifier', SVC())
])

# Setup grid
param_grid = {
    'scaler': [StandardScaler(), RobustScaler(), None],
    'classifier__C': [0.01, 0.1, 1, 10, 100],
    'classifier__gamma': [0.001, 0.01, 0.1, 1],
    'classifier__kernel': ['rbf']
}

# Perform grid search
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy', return_train_score=True)
grid.fit(X_train, y_train)


Training set: 1073 samples, Testing set: 461 samples


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('classifier', SVC())]),
             param_grid={'classifier__C': [0.01, 0.1, 1, 10, 100],
                         'classifier__gamma': [0.001, 0.01, 0.1, 1],
                         'classifier__kernel': ['rbf'],
                         'scaler': [StandardScaler(), RobustScaler(), None]},
             return_train_score=True, scoring='accuracy')

___

#### 2.1.3 - Visualize Grid Search Results

Print the results from the grid search. The following outputs must be printed.
* Best parameter set
* Best cross-validation train score
* Best cross-validation test score
* Test set accuracy

In [24]:
# TODO: Print the results from the grid search

print("Best params:\n{}\n".format(grid.best_params_))
print("Best cross-validation train score: {:.2f}".format(max(grid.cv_results_['mean_train_score'])))
print("Best cross-validation validation score: {:.2f}".format(grid.best_score_))

test_set_score = accuracy_score(y_test, grid.best_estimator_.predict(X_test))
print("Test-set score: {:.2f}".format(test_set_score))

Best params:
{'classifier__C': 100, 'classifier__gamma': 0.01, 'classifier__kernel': 'rbf', 'scaler': RobustScaler()}

Best cross-validation train score: 0.97
Best cross-validation validation score: 0.82
Test-set score: 0.83


Display the confusion matrix and print the classification report for the trained model using the *test* split. 

In [25]:
# TODO - Display confusion matrix and print classification reports
y_pred = grid.best_estimator_.predict(X_test)

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Confusion Matrix:
[[154   1   0]
 [ 20 173  14]
 [  1  44  54]]

Classification Report:
              precision    recall  f1-score   support

        rest       0.88      0.99      0.93       155
    straight       0.79      0.84      0.81       207
        turn       0.79      0.55      0.65        99

    accuracy                           0.83       461
   macro avg       0.82      0.79      0.80       461
weighted avg       0.82      0.83      0.82       461



___

#### 2.1.4 - Question


1. Consider the differences between the performance on the training/cross-validation dataset for section 1 (1.2.3) and section 2 (2.1.3). Which model is a better fit based on *just these two evaluations*? (4 marks)

#### 2.1.4 Answer 
1. Differences:
    - **Section 1:** Training Score: 1.00, Validation Score: 0.95, Test Accuracy: 0.96
    - **Section 2:** Training Score: 0.97, Validation Score: 0.82, Test Accuracy: 0.83
    - **Confusion Matrix and Classification Report:** Section 1 shows near-perfect classification across all classes, while Section 2 struggles, particularly with the "turn" class, which has a low recall (55%).

    Section 1 demonstrates strong performance on both training and test data, suggesting better generalization.
    Section 2 has a larger gap between training (0.97) and validation/test performance (0.82/0.83), indicating some difficulty generalizing to unseen data.

    Therefore, the model in section 1 (1.2.3) is a better fit, because it demonstrates a higher test accuracy (0.96 vs. 0.83), stronger generalization to unseen data, and balanced performance across all motion states.

___

### 2.2 - Model Evaluation on Test Datasets 1 and 2 (7 Marks)

___

#### 2.2.1 - Extract selected features and targets from test datasets

The model trained in 2.1 is to be tested on the independent datasets, that is test_set1 and test_set2. 
Seperate the testsets into the selected features and target vectors *(X1, y1)* and (X2,)

In [26]:
# TODO - Load the target vector from test_set1 into y1, and the selected features from test_set1 into X1
# (1 marks)
selected_features = ['ax_var', 'ay_var', 'az_sum', 'a_sum', 'gz_var']
y1 = test_set1['motion_state']
X1 = test_set1[selected_features]
X1.head(2)

ax_var        ay_var    az_sum     a_sum        gz_var
0  0.000001  7.749919e-08  9.786930  9.801661  7.727190e-10
1  0.000001  9.243000e-08  9.788127  9.802481  8.871433e-10

In [27]:
# TODO - Load the target vector from test_set2 into y2, and the selected features from test_set2 into X2
y2 = test_set2['motion_state']
X2 = test_set2[selected_features]
X2.head(2)

ax_var        ay_var    az_sum     a_sum        gz_var
0  0.000002  3.020959e-07  9.791451  9.800683  1.405038e-09
1  0.000003  6.859937e-07  9.785619  9.797741  1.164381e-09

___

#### 2.2.2 - Visualize Results

For each test dataset, print the test set accuracy score. 

In [28]:
# TODO - Print test score accuracies for test_set1 and test_set2

test_set1_score = accuracy_score(y1, grid.best_estimator_.predict(X1))
test_set2_score = accuracy_score(y2, grid.best_estimator_.predict(X2))

print("test_set1 score: {:.2f}".format(test_set1_score))
print("test_set2 score: {:.2f}".format(test_set2_score))

test_set1 score: 0.83
test_set2 score: 0.84


Display the confusion matrix and print the classification for test_set 1 *(X1, y1)*

In [29]:
# TODO - Display confusion matrix and print classification report for (X1, y1)
y1_pred = grid.best_estimator_.predict(X1)
print("\nConfusion Matrix for test_set1:")
print(confusion_matrix(y1, y1_pred))

print("\nClassification Report for test_set1:")
print(classification_report(y1, y1_pred))


Confusion Matrix for test_set1:
[[229   2   0]
 [ 26 336  49]
 [  0  57  84]]

Classification Report for test_set1:
              precision    recall  f1-score   support

        rest       0.90      0.99      0.94       231
    straight       0.85      0.82      0.83       411
        turn       0.63      0.60      0.61       141

    accuracy                           0.83       783
   macro avg       0.79      0.80      0.80       783
weighted avg       0.83      0.83      0.83       783



Display the confusion matrix and print the classification for test_set 2 *(X2, y2)*

In [30]:
# TODO - Display confusion matrix and print classification report for (X2, y2)
y2_pred = grid.best_estimator_.predict(X2)
print("\nConfusion Matrix for test_set2:")
print(confusion_matrix(y2, y2_pred))

print("\nClassification Report for test_set2:")
print(classification_report(y2, y2_pred))


Confusion Matrix for test_set2:
[[ 15   1   0]
 [  0 178  28]
 [  0  18  57]]

Classification Report for test_set2:
              precision    recall  f1-score   support

        rest       1.00      0.94      0.97        16
    straight       0.90      0.86      0.88       206
        turn       0.67      0.76      0.71        75

    accuracy                           0.84       297
   macro avg       0.86      0.85      0.85       297
weighted avg       0.85      0.84      0.84       297



___

#### 2.2.3 Questions

1. For section 2, compare between the model's performance on the training/validation (2.1) set to the the performance on the two testing sets (2.2). How does the model fit? (3 marks)
1. Compare the models' performances and fits between section 1 and section 2, and consider all three datasets. (3 marks)

#### 2.2.3 Answers
1. Performance Comparison 2.1 vs 2.2

    **Training/Validation Performance (2.1):**
    - Training Accuracy: 0.97
    - Validation Accuracy: 0.82
    - Class Imbalances: High performance on "rest" and "straight" but struggles with "turn" (recall = 55%).

    **Test Set 1 Performance (2.2):**
    - Test Set 1 Accuracy: 0.83
    - Precision and recall for "turn" are lower compared to "rest" and "straight."

    **Test Set 2 Performance (2.2):**
    - Test Set 2 Accuracy: 0.84
    - Performance on "turn" is still weaker (precision = 0.67, recall = 0.76) compared to "rest" and "straight."

    The model fits reasonably well, showing consistent generalization from training/validation to the test datasets. It achieves similar validation and test scores, which suggests it is neither overfitting nor underfitting. However, it consistently struggles with the "turn" class, showing that its performance is limited by the reduced feature set.

2. Performance Comparison Section 1 vs Section 2

    **Training/Validation:**
    - Section 1: Training Score: 1.00, Validation Score: 0.95
    - Section 2: Training Score: 0.97, Validation Score: 0.82

    **Test Set 1:**
    - Section 1: Test Set 1 Accuracy: 0.92, strong precision, recall, and F1-scores across all classes, "Turn" class performs particularly well (precision = 0.89, recall = 0.95).
    - Section 2: Test Set 1 Accuracy: 0.83 (lower than Section 1), struggles with the "turn" class (precision = 0.63, recall = 0.60).

    **Test Set 2:**
    - Section 1: Test Set 2 Accuracy: 0.25 (very poor performance due to complete misclassification of "rest" and "straight").
    - Section 2: Test Set 2 Accuracy: 0.84 (maintains good precision and recall for "rest" and "straight": precision = 0.90+).

    In summary, Section 1 model excels in training, validation, and Test_set1, showing better performance when the test data resembles the training data. Section 2 model is weaker on training, validation, and Test_set1, but significantly outperforms Section 1 on Test_set2, suggesting it generalizes better when faced with data shift.

____

## Section 3 - Classification Using *Group Selected Features* From Inertial Sensor Data (10 Marks)

Once again, an *SVC* model is to be trained and tuned using grid search. However, you must select which features to use from the ones provided in *cv_set*.  

____

### 3.1 Grid Search with Cross Validation (4 Marks)

___

#### 3.1.1 Extract *Group Selected* features and target from *cv_set*

**Question**
1. Please list the features that you have selected and explain *why you have chosen* these features? (4 marks)


**Answer**
1.  We chose the following features: ax_var, ay_var, az_sum, g_sum, and a_var.

    The reason why we chose these features because I wanted a variety of features representing different aspects of motion:
    - Linear Motion: ax_var, ay_var, az_sum
    - Gyoscopic Activity: g_sum
    - Motion Variance: a_var
 
    We wanted to avoid adding too many features compared to Section 1 to prevent overfitting, and we also wanted to capture different features compared to Section 2 that may be more relevant, in order to see the effects on performance across datasets.

In [55]:
# TODO - Load group selected features from cv_set to X
selected_features = ['ax_var', 'ay_var', 'az_sum', 'g_sum', 'a_var']
X = cv_set[selected_features]
X.head(2)

ax_var        ay_var    az_sum     g_sum         a_var
0  8.046094e-07  1.319665e-07  9.789349  0.022554  3.334598e-08
1  1.032619e-06  1.073067e-07  9.789095  0.021731  1.588712e-08

In [ ]:
# TODO - Load motion_state column (target vector) into y
y = cv_set['motion_state']
y.head(2)

0    rest
1    rest
Name: motion_state, dtype: object

___

#### 3.1.2 - Configure and Perform Grid Search with Cross-Validation

Follow the steps from **1.2.2** and perform model training and tuning using the group selected feature set.

In [57]:
# TODO - Perform model training and tuning using `GridSearchCV`

# Perform test/train split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.30, random_state=42, stratify=y
)
print(f"Training set: {X_train.shape[0]} samples, Testing set: {X_test.shape[0]} samples")

# Initialize Pipeline into variable pipe
pipe = Pipeline([
    ('scaler', StandardScaler()),  # Default scaler
    ('classifier', SVC())  # Default classifier
])

# Setup grid set
param_grid = {
    'scaler': [StandardScaler(), RobustScaler(), None],
    'classifier__C': [0.01, 0.1, 1, 10, 100],
    'classifier__gamma': [0.001, 0.01, 0.1, 1],
    'classifier__kernel': ['rbf']  # Radial Basis Function
}

# Perform the grid search to train model and tune hyperparameters
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy', return_train_score=True)
grid.fit(X_train, y_train)

Training set: 1073 samples, Testing set: 461 samples


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('classifier', SVC())]),
             param_grid={'classifier__C': [0.01, 0.1, 1, 10, 100],
                         'classifier__gamma': [0.001, 0.01, 0.1, 1],
                         'classifier__kernel': ['rbf'],
                         'scaler': [StandardScaler(), RobustScaler(), None]},
             return_train_score=True, scoring='accuracy')

___

#### 3.1.3 - Visualize Grid Search Results

Print the results from the grid search. The following outputs must be printed.
* Best parameter set
* Best cross-validation train score
* Best cross-validation test score
* Test set accuracy

In [58]:
# TODO: Print the results from the grid search

print("Best params:\n{}\n".format(grid.best_params_))
print("Best cross-validation train score: {:.2f}".format(max(grid.cv_results_['mean_train_score'])))
print("Best cross-validation validation score: {:.2f}".format(grid.best_score_))

test_set_score = accuracy_score(y_test, grid.best_estimator_.predict(X_test))
print("Test-set score: {:.2f}".format(test_set_score))

Best params:
{'classifier__C': 100, 'classifier__gamma': 0.01, 'classifier__kernel': 'rbf', 'scaler': RobustScaler()}

Best cross-validation train score: 0.97
Best cross-validation validation score: 0.90
Test-set score: 0.93


Display the confusion matrix and print the classification report for the trained model using the *test* split. 

In [59]:
# TODO - Display confusion matrix and print classification report
y_pred = grid.best_estimator_.predict(X_test)

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Confusion Matrix:
[[154   1   0]
 [ 20 184   3]
 [  0   9  90]]

Classification Report:
              precision    recall  f1-score   support

        rest       0.89      0.99      0.94       155
    straight       0.95      0.89      0.92       207
        turn       0.97      0.91      0.94        99

    accuracy                           0.93       461
   macro avg       0.93      0.93      0.93       461
weighted avg       0.93      0.93      0.93       461



___

### 3.2 - Model Evaluation on Test Datasets 1 and 2 (6 Marks)

___

#### 3.2.1 - Extract group selected features and targets from test datasets

The model trained in 3.1 is to be tested on the independent datasets, that is test_set1 and test_set2. 
Seperate the testsets into group selected features and target vectors *(X1, y1)* and *(X2,y2)*

In [60]:
# TODO - Load the target vector from test_set1 into y1, and the group selected features from test_set1 into X1
selected_features = ['ax_var', 'ay_var', 'az_sum', 'g_sum', 'a_var']
y1 = test_set1['motion_state']
X1 = test_set1[selected_features]
X1.head(2)

ax_var        ay_var    az_sum     g_sum         a_var
0  0.000001  7.749919e-08  9.786930  0.025634  3.437614e-08
1  0.000001  9.243000e-08  9.788127  0.025218  3.613404e-08

In [61]:
# TODO - Load the target vector from test_set2 into y2, and the group selected features from test_set2 into X2
y2 = test_set2['motion_state']
X2 = test_set2[selected_features]
X2.head(2)

ax_var        ay_var    az_sum     g_sum         a_var
0  0.000002  3.020959e-07  9.791451  0.012437  6.614559e-08
1  0.000003  6.859937e-07  9.785619  0.013805  3.344639e-07

___

#### 3.2.2 - Visualize Results

For each test dataset, print the test set accuracy score. 

In [62]:
# TODO - Print test score accuracies for test_set1 and test_set2
test_set1_score = accuracy_score(y1, grid.best_estimator_.predict(X1))
test_set2_score = accuracy_score(y2, grid.best_estimator_.predict(X2))

print("test_set1 score: {:.2f}".format(test_set1_score))
print("test_set2 score: {:.2f}".format(test_set2_score))

test_set1 score: 0.80
test_set2 score: 0.88


Display the confusion matrix and print the classification for test_set 1 *(X1, y1)*

In [63]:
# TODO - Display confusion matrix and print classification report for (X1, y1)
y1_pred = grid.best_estimator_.predict(X1)
print("\nConfusion Matrix for test_set1:")
print(confusion_matrix(y1, y1_pred))

print("\nClassification Report for test_set1:")
print(classification_report(y1, y1_pred))


Confusion Matrix for test_set1:
[[141  90   0]
 [ 12 377  22]
 [  0  33 108]]

Classification Report for test_set1:
              precision    recall  f1-score   support

        rest       0.92      0.61      0.73       231
    straight       0.75      0.92      0.83       411
        turn       0.83      0.77      0.80       141

    accuracy                           0.80       783
   macro avg       0.84      0.76      0.79       783
weighted avg       0.82      0.80      0.79       783



Display the confusion matrix and print the classification for test_set 2 *(X2, y2)*

In [64]:
# TODO - Display confusion matrix and print classification report for (X2, y2)
y2_pred = grid.best_estimator_.predict(X2)
print("\nConfusion Matrix for test_set2:")
print(confusion_matrix(y2, y2_pred))

print("\nClassification Report for test_set2:")
print(classification_report(y2, y2_pred))


Confusion Matrix for test_set2:
[[  6  10   0]
 [  0 202   4]
 [  0  21  54]]

Classification Report for test_set2:
              precision    recall  f1-score   support

        rest       1.00      0.38      0.55        16
    straight       0.87      0.98      0.92       206
        turn       0.93      0.72      0.81        75

    accuracy                           0.88       297
   macro avg       0.93      0.69      0.76       297
weighted avg       0.89      0.88      0.87       297



___

#### 3.2.3 Questions

1. Based on the observations across all three datasets, compare the performance of the model trained in this section to: (3 marks)
    1. Model trained in Section 1 using *all features*.
    1. Model trained in Section 2 using *selected features*. 

2. Please proivde potential explanations for the differences observed in models' performances across the three sections. (3 marks)

#### 3.2.3 Answers
1. Performance Comparison
    1. Model Section 1 vs Section 3
        
        **Training/Validation Data:**
        - Validation Accuracy: Section 1 (0.95) was greater than Section 3 (0.90)
        - Test Accuracy: Section 1 (0.96) was greater than Section 3 (0.93)

        **Test Set 1:**
        - Section 1 had a higher accuracy of 0.92 compared to Section 3 with an accuracy of 0.80
        - Section 1 had a strong classification of "rest" and "straight," with minor errors in "turn."
        - Section 3 struggled with "rest" (precision = 0.92, recall = 0.61) 

        **Test Set 2:**
        - Section 3 had a significantly higher accuracy of 0.88 compared to Section 1 with an accuracy of 0.25
        - Section 1 was unable to classify "rest" and "straight."
        - Section 3 had high precision for "straight" and "turn" classes.

    2. Model Section 2 vs Section 3

        **Training/Validation Data:**
        - Validation Accuracy: Section 3 (0.90) was greater than Section 2 (0.82)
        - Test Accuracy: Section 3 (0.93) was greater than Section 2 (0.83)

        **Test Set 1:**
        - Section 2 had a higher accuracy of 0.83 compared to Section 3 with an accuracy of 0.80
        - Section 2 struggled with "turn" (precision = 0.63, recall = 0.60).

        **Test Set 2:**
        - Section 3 had a higher accuracy of 0.88 compared to Section 2 with an accuracy of 0.84
        - Both had strong performance across all classes

2. Potential Explanations for the Differences
    1. **Feature Set Size:** 
        - Section 1 benefits from using all features, leading to better performance on similar datasets like Test Set 1, but it overfits to the training distribution, causing poor generalization to Test Set 2. 
        - Section 2 reduces features significantly, leading to weaker performance on similar datasets but better adaptability to Test Set 2. The features in Section 2 appear to align better with the characteristics of Test Set 1 compared to Section 3. 
        - Section 3 appears to have a good balance of features with solid overall performance across all datasets using the group-selected features.

    2. **Overfitting**
        - Section 1 shows signs of overfitting, with great training scores and poor performance on Test Set 2
        - Sections 2 and 3 show better generalization on Test Set 2

    3. **Data Shift in Test Set 2**
        - Test Set 2 has different feature distributions compared to training data and Test Set 1, explaining the difference in performance.
        - Section 2 and 3 both adapt better to these shifts, with Section 3 performing the best.

____

## Section 4 - Classification Using Principal Components of Features From Inertial Sensor Data (10 Marks)

____

### 4.1 Principal Component Analysis (PCA) Search with Cross Validation (4 Marks)

___

#### 4.1.1 Extract *ALL* features and target from *cv_set*

In [79]:
# TODO - Load all features from cv_set to X
X = cv_set.drop(columns=['motion_state'])
X.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0 -0.003929 -0.093081  9.789349 -0.002649  0.001406 -0.001391  9.797418   
1 -0.000966 -0.092417  9.789095 -0.001806  0.000103 -0.001943  9.798815   

      g_sum        ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.022554  8.046094e-07  1.319665e-07  ...  0.023871  0.000003  0.001496   
1  0.021731  1.032619e-06  1.073067e-07  ...  0.024153  0.000006  0.001865   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.000544  0.024992  0.000110  0.000075  0.000078  0.024997  0.000187  
1  0.000615  0.024828  0.000124  0.000098  0.000090  0.024829  0.000144  

[2 rows x 32 columns]

In [80]:
# TODO - Load motion_state column (target vector) into y
y = cv_set['motion_state']
y.head(2)

0    rest
1    rest
Name: motion_state, dtype: object

___

#### 4.1.2 - Configure and Perform Grid Search with Cross-Validation

Similar to the steps from **1.2.2** and perform model training and tuning for an `SVC` model. Perform the `train_test_split`, initiate a `Pipeline`, then define the *Parameters* for `GridSearchCV`. Finally, perfrom the grid search.

In [81]:
# TODO - Perform model training and tuning using `GridSearchCV`
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.30, random_state=42, stratify=y
)
print(f"Training set: {X_train.shape[0]} samples, Testing set: {X_test.shape[0]} samples")

Training set: 1073 samples, Testing set: 461 samples


In the `Pipeline` add an entry *'pca'* to apply `PCA`.

In [82]:
# Perform any missing imports
from sklearn.decomposition import PCA

# TODO - Initiate the pipeline 
# (1 marks)
pipe = Pipeline([
    ('scaler', StandardScaler()),  # Default scaler
    ('pca', PCA()),  # Principal Component Analysis
    ('classifier', SVC())  # Support Vector Classifier
])

Similar to previous sections, define the grid search configuration (*param_grid*). Add *n_components* hyper parameter tuning for PCA, and search over the values `n_components=[4,7,11,14,16]`.  

In [83]:
# TODO - Define the parameter grid for your model
# (3 marks)
param_grid = {
    'scaler': [StandardScaler(), RobustScaler(), None],
    'pca__n_components': [4, 7, 11, 14, 16],  # Number of principal components
    'classifier__C': [0.01, 0.1, 1, 10, 100],
    'classifier__gamma': [0.001, 0.01, 0.1, 1],
    'classifier__kernel': ['rbf']  # Radial Basis Function kernel
}

In [84]:
# Perform the grid search to train model and tune hyper parameters
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy', return_train_score=True)
grid.fit(X_train, y_train) 

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('pca', PCA()), ('classifier', SVC())]),
             param_grid={'classifier__C': [0.01, 0.1, 1, 10, 100],
                         'classifier__gamma': [0.001, 0.01, 0.1, 1],
                         'classifier__kernel': ['rbf'],
                         'pca__n_components': [4, 7, 11, 14, 16],
                         'scaler': [StandardScaler(), RobustScaler(), None]},
             return_train_score=True, scoring='accuracy')

___

#### 4.1.3 - Visualize Grid Search Results

Print the results from the grid search. The following outputs must be printed.
* Best parameter set
* Best cross-validation train score
* Best cross-validation test score
* Test set accuracy

In [85]:
# TODO: Print the results from the grid search
print("Best params:\n{}\n".format(grid.best_params_))
print("Best cross-validation train score: {:.2f}".format(max(grid.cv_results_['mean_train_score'])))
print("Best cross-validation validation score: {:.2f}".format(grid.best_score_))

test_set_score = accuracy_score(y_test, grid.best_estimator_.predict(X_test))
print("Test-set score: {:.2f}".format(test_set_score))

Best params:
{'classifier__C': 100, 'classifier__gamma': 0.01, 'classifier__kernel': 'rbf', 'pca__n_components': 14, 'scaler': RobustScaler()}

Best cross-validation train score: 1.00
Best cross-validation validation score: 0.94
Test-set score: 0.96


Display the confusion matrix and print the classification report for the trained model using the *test* split. 

In [86]:
# TODO - Display confusion matrix and print classification report
y_pred = grid.best_estimator_.predict(X_test)

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Confusion Matrix:
[[154   1   0]
 [  9 192   6]
 [  0   2  97]]

Classification Report:
              precision    recall  f1-score   support

        rest       0.94      0.99      0.97       155
    straight       0.98      0.93      0.96       207
        turn       0.94      0.98      0.96        99

    accuracy                           0.96       461
   macro avg       0.96      0.97      0.96       461
weighted avg       0.96      0.96      0.96       461



___

### 4.2 - Model Evaluation on Test Datasets 1 and 2 (6 Marks)

___

#### 4.2.1 - Extract group selected features and targets from test datasets

In [87]:
# TODO - Load the target vector from test_set1 into y1, and all features from test_set1 into X1
y1 = test_set1['motion_state']
X1 = test_set1.drop(columns=['motion_state'])
X1.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0  0.264465  0.089904  9.786930 -0.002305  0.000218 -0.001143  9.801661   
1  0.264300  0.088882  9.788127 -0.001865 -0.000523 -0.001999  9.802481   

      g_sum    ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.025634  0.000001  7.749919e-08  ...  0.024056  0.000003  0.002506   
1  0.025218  0.000001  9.243000e-08  ...  0.024084  0.000007  0.002427   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.000845  0.024884  0.000155  0.000078  0.000069  0.024886  0.000170  
1  0.000949  0.024881  0.000201  0.000095  0.000076  0.024883  0.000201  

[2 rows x 32 columns]

In [88]:
# TODO - Load the target vector from test_set2 into y2, and all features from test_set2 into X2
y2 = test_set2['motion_state']
X2 = test_set2.drop(columns=['motion_state'])
X2.head(2)

ax_sum    ay_sum    az_sum    gx_sum    gy_sum    gz_sum     a_sum  \
0 -0.046400  0.305907  9.791451 -0.002184  0.001311 -0.001901  9.800683   
1 -0.050073  0.311947  9.785619 -0.003369  0.001898 -0.002051  9.797741   

      g_sum    ax_var        ay_var  ...     a_min     g_min    ax_max  \
0  0.012437  0.000002  3.020959e-07  ...  0.048406  0.000007  0.002575   
1  0.013805  0.000003  6.859937e-07  ...  0.047696  0.000013  0.003880   

     ay_max    az_max    gx_max    gy_max    gz_max     a_max     g_max  
0  0.002993  0.049701  0.000094  0.000115  0.000080  0.049764  0.000142  
1  0.004106  0.050477  0.000140  0.000140  0.000083  0.050763  0.000151  

[2 rows x 32 columns]

___

#### 4.2.2 - Visualize Results

In [89]:
# TODO - Print test score accuracies for test_set1 and test_set2
test_set1_score = accuracy_score(y1, grid.best_estimator_.predict(X1))
test_set2_score = accuracy_score(y2, grid.best_estimator_.predict(X2))

print("test_set1 score: {:.2f}".format(test_set1_score))
print("test_set2 score: {:.2f}".format(test_set2_score))

test_set1 score: 0.93
test_set2 score: 0.67


In [90]:
# TODO - Display confusion matrix and print classification report for (X1, y1)
y1_pred = grid.best_estimator_.predict(X1)
print("\nConfusion Matrix for test_set1:")
print(confusion_matrix(y1, y1_pred))

print("\nClassification Report for test_set1:")
print(classification_report(y1, y1_pred))


Confusion Matrix for test_set1:
[[223   8   0]
 [ 20 372  19]
 [  0   6 135]]

Classification Report for test_set1:
              precision    recall  f1-score   support

        rest       0.92      0.97      0.94       231
    straight       0.96      0.91      0.93       411
        turn       0.88      0.96      0.92       141

    accuracy                           0.93       783
   macro avg       0.92      0.94      0.93       783
weighted avg       0.93      0.93      0.93       783



Display the confusion matrix and print the classification for test_set 2 *(X2, y2)*

In [91]:
# TODO - Display confusion matrix and print classification report for (X2, y2)
print("\nConfusion Matrix for test_set2:")
print(confusion_matrix(y2, y2_pred))

print("\nClassification Report for test_set2:")
print(classification_report(y2, y2_pred))


Confusion Matrix for test_set2:
[[  6  10   0]
 [  0 202   4]
 [  0  21  54]]

Classification Report for test_set2:
              precision    recall  f1-score   support

        rest       1.00      0.38      0.55        16
    straight       0.87      0.98      0.92       206
        turn       0.93      0.72      0.81        75

    accuracy                           0.88       297
   macro avg       0.93      0.69      0.76       297
weighted avg       0.89      0.88      0.87       297



___

#### 4.2.3 - Questions

1. Based on the observations across all three datasets, compare the performance of the model with PCA, trained in this section to: (4 marks)
    * Model trained in Section 1 using *all features*.
    * Model trained in Section 2 using *selected features*. 
    * Model trained in Section 3 using *group selected features*

2. Please provide potential explanations for the differences observed in models' performances across between section 1 and 4. (2 marks)

#### 4.2.3 - Answers
1. Performance Comparison
    1. Model Section 1 vs Section 4
        
        **Training/Validation Data:**
        - Validation Accuracy: Section 1 (0.95) is slightly greater than Section 4 (0.94)
        - Test Accuracy: Section 1 (0.96) is equal to Section 3 (0.96)

        **Test Set 1:**
        - Section 4 had a slightly higher accuracy of 0.93 compared to Section 1 with an accuracy of 0.92
        - Section 4 achieves slightly higher accuracy and stronger performance for "turn".

        **Test Set 2:**
        - Section 4 had a significantly higher accuracy of 0.67 compared to Section 1 with an accuracy of 0.25

    2. Model Section 2 vs Section 4

        **Training/Validation Data:**
        - Validation Accuracy: Section 4 (0.94) was greater than Section 2 (0.82)
        - Test Accuracy: Section 4 (0.96) was greater than Section 2 (0.83)

        **Test Set 1:**
        - Section 4 had a higher accuracy of 0.93 compared to Section 2 with an accuracy of 0.83

        **Test Set 2:**
        - Section 2 had a higher accuracy of 0.84 compared to Section 4 with an accuracy of 0.67

    3. Model Section 3 vs Section 4

        **Training/Validation Data:**
        - Validation Accuracy: Section 4 (0.94) was slightly greater than Section 3 (0.90)
        - Test Accuracy: Section 4 (0.96) was slightly greater than Section 2 (0.93)

        **Test Set 1:**
        - Section 4 had a higher accuracy of 0.93 compared to Section 2 with an accuracy of 0.80

        **Test Set 2:**
        - Section 2 had a higher accuracy of 0.88 compared to Section 4 with an accuracy of 0.67

2. Potential Explanations for the Differences
    1. **Dimensionality Reduction:** 
        - Section 4 uses PCA to focus on the most important components of variability in the data, reducing noise and redundant features, these leads to Section 4 outperforming Section 1 in test sets.
        - Section 2 reduces features significantly, leading to weaker performance on similar datasets but better adaptability to Test Set 2. The features in Section 2 appear to align better with the characteristics of Test Set 1 compared to Section 3. 
        - Section 1 uses all features, which can include irrelevant or noisy data, leading to overfitting on similar datasets and poor generalization on shifted datasets like Test Set 2

    2. **Overfitting**
        - PCA in Section 4 enhances generalization under data shifts (Test Set 2) compared to overfitting in Section 1.
        - Section 1 achieves perfect training scores (1.00) and high validation accuracy but overfits the training data
        - Sections 2 and 3 still show better generalization on Test Set 2 compared to Section 1 and 4

    3. **Feature Representation**
        - Section 4's PCA balances representation by reducing dimensionality while maintaining critical information. This leads to strong performance on Test Set 1 (0.93 accuracy) and moderate generalization to Test Set 2, but it may lose some important details critical for specific distributions.
        - As shown for Test Set 2, PCA in Section 4 might discard some information critical for Test Set 2's distribution which results in a weaker performance compared to Section 2 and Section 3

____

## Section 5 - Classification Using *Pre Selected Features* Derived From Inertial Sensor Data (10 Marks)

___

Now, a second classifer is to be trained and tuned using grid search. The **selected features from section 2** must be used. You may choose from one of the algorithms discussed that class listed below.
1. Random Forest
1. K-Nearest Neighbors
1. Gradient Boosting Machines
1. Neural Networks

____

### 5.1 Grid Search with Cross Validation (6 Marks)

___

#### 5.1.1 Extract *Selected* features and target from *cv_set*

**Questions**
1. Please list the ML model that you have selected . And explain *why you have chosen* this model? (2 marks)

**Answer** 
1. We selected a Random Forest. The reason why we picked Random Forest is because it handles overfitting well. Also, the features we chose in section 2, might not all be equally important. Random Forest is good at figuring out which features matter more and uses that information to make better predictions. Also Random Forest usually works well on different types of problems without needing a lot of extra work, like scaling or changing the data. The other methods may take a longer time to calculate distances if the dataset is big, or they may need more careful tuning and more time to train.

In [92]:
# TODO - Load group selected features from cv_set to X
selected_features = ['ax_var', 'ay_var', 'az_sum', 'a_sum', 'gz_var']
X = cv_set[selected_features]
X.head(2)

ax_var        ay_var    az_sum     a_sum        gz_var
0  8.046094e-07  1.319665e-07  9.789349  9.797418  7.446916e-10
1  1.032619e-06  1.073067e-07  9.789095  9.798815  6.983481e-10

In [93]:
# TODO - Load motion_state column (target vector) into y
y = cv_set['motion_state']
y.head(2)

0    rest
1    rest
Name: motion_state, dtype: object

___

#### 5.1.2 - Configure and Perform Grid Search with Cross-Validation

Similar to the steps from **1.2.2** and perform model training and tuning using for the model selected. Perform the `train_test_split`, initiate a `Pipeline`, then define the *Parameters* for `GridSearchCV`. Finally, perfrom the grid search.

In [94]:
# TODO - Perform train/test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.30, random_state=42, stratify=y
)
print(f"Training set: {X_train.shape[0]} samples, Testing set: {X_test.shape[0]} samples")

Training set: 1073 samples, Testing set: 461 samples


In [95]:
# Perform any missing imports
from sklearn.ensemble import RandomForestClassifier

# TODO - Initiate the pipeline 
# (1 mark)
pipe = Pipeline([
    ('classifier', RandomForestClassifier(random_state=42))
])

In [96]:
# TODO - Define the parameter grid for your model
# (3 marks)
param_grid = {
    'classifier__n_estimators': [50, 100, 200],  # Number of trees in the forest
    'classifier__max_depth': [None, 10, 20, 30],  # Maximum depth of each tree
    'classifier__min_samples_split': [2, 5, 10],  # Minimum samples required to split an internal node
    'classifier__min_samples_leaf': [1, 2, 4],  # Minimum samples required to be at a leaf node
    'classifier__bootstrap': [True, False]  # Whether bootstrap samples are used when building trees
}

In [ ]:
# TODO - Perform the grid search
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='accuracy', return_train_score=True)
grid.fit(X_train, y_train) # Took approximately 2m 30.4s to execute

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('classifier',
                                        RandomForestClassifier(random_state=42))]),
             param_grid={'classifier__bootstrap': [True, False],
                         'classifier__max_depth': [None, 10, 20, 30],
                         'classifier__min_samples_leaf': [1, 2, 4],
                         'classifier__min_samples_split': [2, 5, 10],
                         'classifier__n_estimators': [50, 100, 200]},
             return_train_score=True, scoring='accuracy')

___

#### 5.1.3 - Visualize Grid Search Results

Print the results from the grid search. The following outputs must be printed.
* Best parameter set
* Best cross-validation train score
* Best cross-validation test score
* Test set accuracy

In [98]:
# TODO: Print the results from the grid search

print("Best params:\n{}\n".format(grid.best_params_))
print("Best cross-validation train score: {:.2f}".format(max(grid.cv_results_['mean_train_score'])))
print("Best cross-validation validation score: {:.2f}".format(grid.best_score_))

test_set_score = accuracy_score(y_test, grid.best_estimator_.predict(X_test))
print("Test-set score: {:.2f}".format(test_set_score))

Best params:
{'classifier__bootstrap': True, 'classifier__max_depth': None, 'classifier__min_samples_leaf': 2, 'classifier__min_samples_split': 10, 'classifier__n_estimators': 50}

Best cross-validation train score: 0.99
Best cross-validation validation score: 0.77
Test-set score: 0.77


Display the confusion matrix and print the classification report for the trained model using the *test* split. 

In [99]:
# TODO - Display confusion matrix and print classification report
y_pred = grid.best_estimator_.predict(X_test)

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Confusion Matrix:
[[142  12   1]
 [  9 185  13]
 [  1  68  30]]

Classification Report:
              precision    recall  f1-score   support

        rest       0.93      0.92      0.93       155
    straight       0.70      0.89      0.78       207
        turn       0.68      0.30      0.42        99

    accuracy                           0.77       461
   macro avg       0.77      0.70      0.71       461
weighted avg       0.77      0.77      0.75       461



___

### 5.2 - Model Evaluation on Test Datasets 1 and 2 (4 Marks)

___

#### 5.2.1 - Extract selected features and targets from test datasets

The model trained in 5.1, is to be tested on the independent datasets, that is test_set1 and test_set2. 
Seperate the testsets into features and target vectors *(X1, y1)* and *(X2,y2)*

In [100]:
# TODO - Load the target vector from test_set1 into y1, and the selected features from test_set1 into X1
selected_features = ['ax_var', 'ay_var', 'az_sum', 'a_sum', 'gz_var']
y1 = test_set1['motion_state']
X1 = test_set1[selected_features]
X1.head(2)

ax_var        ay_var    az_sum     a_sum        gz_var
0  0.000001  7.749919e-08  9.786930  9.801661  7.727190e-10
1  0.000001  9.243000e-08  9.788127  9.802481  8.871433e-10

In [101]:
# TODO - Load the target vector from test_set2 into y2, and the selected features from test_set2 into X2
y2 = test_set2['motion_state']
X2 = test_set2[selected_features]
X2.head(2)

ax_var        ay_var    az_sum     a_sum        gz_var
0  0.000002  3.020959e-07  9.791451  9.800683  1.405038e-09
1  0.000003  6.859937e-07  9.785619  9.797741  1.164381e-09

___

#### 5.2.2 - Visualize Results

For each test dataset, print the test set accuracy score. 

In [102]:
# TODO - Print test score accuracies for test_set1 and test_set2
test_set1_score = accuracy_score(y1, grid.best_estimator_.predict(X1))
test_set2_score = accuracy_score(y2, grid.best_estimator_.predict(X2))

print("test_set1 score: {:.2f}".format(test_set1_score))
print("test_set2 score: {:.2f}".format(test_set2_score))

test_set1 score: 0.78
test_set2 score: 0.74


Display the confusion matrix and print the classification for test_set 1 *(X1, y1)*

In [103]:
# TODO - Display confusion matrix and print classification report for (X1, y1)
y1_pred = grid.best_estimator_.predict(X1)
print("\nConfusion Matrix for test_set1:")
print(confusion_matrix(y1, y1_pred))

print("\nClassification Report for test_set1:")
print(classification_report(y1, y1_pred))


Confusion Matrix for test_set1:
[[226   5   0]
 [  8 344  59]
 [  0 101  40]]

Classification Report for test_set1:
              precision    recall  f1-score   support

        rest       0.97      0.98      0.97       231
    straight       0.76      0.84      0.80       411
        turn       0.40      0.28      0.33       141

    accuracy                           0.78       783
   macro avg       0.71      0.70      0.70       783
weighted avg       0.76      0.78      0.77       783



Display the confusion matrix and print the classification for test_set 2 *(X2, y2)*

In [104]:
# TODO - Display confusion matrix and print classification report for (X2, y2)
y2_pred = grid.best_estimator_.predict(X2)
print("\nConfusion Matrix for test_set2:")
print(confusion_matrix(y2, y2_pred))

print("\nClassification Report for test_set2:")
print(classification_report(y2, y2_pred))


Confusion Matrix for test_set2:
[[ 13   3   0]
 [  1 200   5]
 [  0  68   7]]

Classification Report for test_set2:
              precision    recall  f1-score   support

        rest       0.93      0.81      0.87        16
    straight       0.74      0.97      0.84       206
        turn       0.58      0.09      0.16        75

    accuracy                           0.74       297
   macro avg       0.75      0.63      0.62       297
weighted avg       0.71      0.74      0.67       297



___

#### 5.2.3 Question

1. For the model trainied in this section, compare its performance to the performance of SVC which was also trained with the same selected features (section 2). Which model fits better? Explain causes for any observed discrepancies? (4 marks)

#### 5.2.3 - Answer 
1. Performance Comparison of Section 2 (SVC) vs Section 5 (Random Forest)

    **Training/Validation Data:**
    - Cross-validation train score: Section 5 (0.99) is slightly greater than Section 2 (0.97)
    - Cross-validation validation score: Section 2 (0.82) is greater than Section 5 (0.77)
    - Test-set score: Section 2 (0.83) is greater than Section 5 (0.77)

    **Test Set 1:**
    - Section 2 had a slightly higher accuracy of 0.83 compared to Section 5 with an accuracy of 0.78
    - Section 5 struggled with "turn" class.

    **Test Set 2:**
    - Section 2 had a higher accuracy of 0.84 compared to Section 5 with an accuracy of 0.74
    - Section 5 continued to struggle with "turn" class.

    SVC (Section 2) fits better overall, as it generalizes well across all datasets and balances performance across classes, especially the minority "turn" class.

    **Reasons for Discrepancies:**
    1. Overfitting:
        - Random Forest has very high training scores but lower validation/test scores, suggesting overfitting to training data. SVC avoids this and generalizes better.

    2. Handling of Features:
        - SVC finds a global boundary using all features, which works well with the selected feature set. Random Forest splits features hierarchically, which can lead to weaker performance with fewer features.

    3. Class Imbalance:
        - The "turn" class has fewer samples. SVC balances class performance better, while Random Forest focuses more on the majority classes ("rest" and "straight"), leading to poor recall for "turn."

    4. Random Forest Complexity:
        - Random Forest has more hyperparameters, making it harder to optimize effectively, whereas SVC is simpler to tune and achieves strong results.

    In summary, the SVC model trained in Section 2 is better suited for the selected features. Random Forest requires further tuning or additional features to match SVC's performance.

____